dist/general_test.go

```diff
@@ -64,126 +61,73 @@ func testDistributionProbs(t *testing.T, dist UniProbDist, name string, pts []un
 }
 
 type ConjugateUpdater interface {
-	MarshalParameters([]Parameter)
-	UnmarshalParameters([]Parameter)
-	Rand() float64
 	NumParameters() int
+	NumSuffStat() int
 	SuffStat([]float64, []float64, []float64) float64
 	ConjugateUpdate([]float64, float64, []float64)
-	NumSuffStat() int
-}
-
-func testConjugateUpdate(t *testing.T, dist ConjugateUpdater, newFittable func() ConjugateUpdater) {
-	samps := make([]float64, 10)
-	for i := range samps {
-		samps[i] = dist.Rand()
-	}
-	nParams := dist.NumParameters()
-	nSuffStat := dist.NumSuffStat()
-
-	stats2 := make([]float64, nSuffStat)
-	p2 := make([]Parameter, nParams)
-	w2 := make([]float64, nParams)
-	n2 := newFittable()
-	ns2 := n2.SuffStat(samps, nil, stats2)
-	n2.ConjugateUpdate(stats2, ns2, w2)
-	n2.MarshalParameters(p2)
-
-	n3 := newFittable()
-	p3 := make([]Parameter, nParams)
-	w3 := make([]float64, nParams)
-	stats3 := make([]float64, nSuffStat)
-	ns3 := n3.SuffStat(samps[:7], nil, stats3)
-	n3.ConjugateUpdate(stats3, ns3, w3)
-	n3.MarshalParameters(p3)
-
-	n4 := newFittable()
-	n4.UnmarshalParameters(p3)
-	p4 := make([]Parameter, nParams)
-	w4 := make([]float64, nParams)
-	stats4 := make([]float64, nSuffStat)
-	copy(w4, w3)
-	ns4 := n4.SuffStat(samps[7:], nil, stats4)
-	n4.ConjugateUpdate(stats4, ns4, w4)
-	n4.MarshalParameters(p4)
-
-	if !parametersEqual(p2, p4, 1e-14) {
-		t.Errorf("prior doesn't match after two step update. First is %v, second is %v", p2, p4)
-	}
-	if !floats.EqualApprox(w2, w4, 1e-14) {
-		t.Errorf("prior weight doesn't match after two step update. First is %v, second is %v", w2, w4)
-	}
-
-	// Try with weights = 1
-	ones := make([]float64, len(samps))
-	for i := range ones {
-		ones[i] = 1
-	}
-
-	n5 := newFittable()
-
-	p5 := make([]Parameter, nParams)
-	w5 := make([]float64, nParams)
-	stats5 := make([]float64, nSuffStat)
-	ns5 := n5.SuffStat(samps, ones, stats5)
-	n5.ConjugateUpdate(stats5, ns5, w5)
-	n5.MarshalParameters(p5)
 
-	if !parametersEqual(p2, p5, 1e-14) {
-		t.Errorf("prior doesn't match after unitary weights. First is %v, second is %v", p2, p5)
-	}
-	if !floats.EqualApprox(w2, w5, 1e-14) {
-		t.Errorf("prior weight doesn't match unitary weights. First is %v, second is %v", w2, w5)
-	}
+	Rand() float64
+}
 
-	// Lastly, make sure it's okay with a bunch of random weights
-	weights := make([]float64, len(samps))
-	for i := range weights {
-		weights[i] = rand.Float64()
+func testConjugateUpdate(t *testing.T, newFittable func() ConjugateUpdater, isEqual func(ConjugateUpdater, ConjugateUpdater) bool) {
+	for i, test := range []struct {
+		samps   []float64
+		weights []float64
+	}{
+		{
```

Comment:
Could you recombine the nil case and the ones case? Maybe with a "if weights == nil, test with ones". Those two answers should be the same, which is otherwise untested. I found bugs in my initial implementations because of checking that case.